clients/src/main/java/org/apache/kafka/common/requests/RequestContext.java

```diff
@@ -41,7 +41,9 @@ public class RequestContext implements AuthorizableRequestContext {
     public final KafkaPrincipal principal;
     public final ListenerName listenerName;
     public final SecurityProtocol securityProtocol;
-    public final ClientInformation clientInformation;
+    // The client information can be updated if the request is ApiVersionRequest,
+    // so the client information will not be unknown for ApiVersionRequest.
+    public ClientInformation clientInformation;
```

Comment:
Making this mutable and public seems dangerous. A few things to consider:
1. Thread safety.
2. Whether we want to allow it to be updated only once (if so, we'd want to have a method that enforces that).